### PR TITLE
Feat CloudWatch metrics recorder

### PR DESCRIPTION
**Pull request by chejimmy**
# Description

- [chore: update @aws-sdk/* to 3.450.0](https://github.com/awslabs/iot-application/pull/1757/commits/a6a16722b0515c728b9e905fc7d7e9cb096ba8a1)
  - update @aws-sdk/* to the latest version that's compatible with `aws-sdk-client-mock`; related https://github.com/m-radzikowski/aws-sdk-client-mock/issues/193
- [feat: cloudwatch metrics recorder implementation](https://github.com/awslabs/iot-application/pull/1757/commits/06a23edb415fc50d44973d0964ba3b0b7779597f)
  - added CloudWatch metrics recorder implementation
  - `metrics-recorder.interface.ts` is copied from [app-kit](https://github.com/awslabs/iot-app-kit/blob/rc/packages/core/src/metric-recorder/metricsRecorder.interface.ts) to unblock implementation until a new version includes the interface is released
- Next action item(s)
  - User permission update
  - Provide the namespace to the CloudWatch metrics recorder

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Test cases added

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Legal

This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

**Review by jmbuss — COMMENTED** [review]
This is sending metrics to the customers account right? Is there some way we can track errors so we know there's any issues for us to fix? or does that require a managed solution?

File: apps/client/src/metrics/cloud-watch-metrics-recorder.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    () => this.putMetrics(),
+    PUT_METRICS_THROTTLE,
+    {
+      leading: false,
+      trailing: true,
+    },
+  );
+
+  private async putMetrics() {
+    const metricData = this.queue.getMetricData();
```
[the remark this answers is not in the document]
Comment (chejimmy) [reply apps/client/src/metrics/cloud-watch-metrics-recorder.ts:100]: Batched under the `MetricData` field https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricData.html

**Comment by chejimmy** [comment]
> This is sending metrics to the customers account right? Is there some way we can track errors so we know there's any issues for us to fix? or does that require a managed solution?

1. Stored under the customer account today
2. Customer can choose to share metrics with us to give us insights.
3. We determined the managed solution was too costly and would not make sense for the scale today.

File: apps/client/src/metrics/cloud-watch-metrics-recorder.spec.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    cloudWatchMetricsRecorder.setMetricNamespace(metricNamespace);
+
+    const timestamp0 = new Date();
+    cloudWatchMetricsRecorder.record({
+      metricName: 'metric0',
+      metricValue: 0,
+    });
+
+    // PUT_METRICS_THROTTLE
+    vi.advanceTimersByTime(3000);
```
Comment (tracy-french) [inline apps/client/src/metrics/cloud-watch-metrics-recorder.spec.ts:146]: Why 3s? Should this be using the constant `PUT_METRICS_THROTTLE`?

File: apps/client/src/metrics/cloud-watch-metrics-recorder.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  getMetricData() {
+    return Array.from(this.queue);
+  }
+
+  clear() {
+    this.queue.clear();
+  }
+
+  private static parseMetric({
```
Comment (tracy-french) [inline apps/client/src/metrics/cloud-watch-metrics-recorder.ts:33]: Why is this method static, if it is private and only referenced within the class by non-static public method?

**Review by tracy-french — APPROVED** [review]

File: apps/client/src/metrics/cloud-watch-metrics-recorder.spec.ts (hunk 2 above)
Comment (chejimmy) [reply apps/client/src/metrics/cloud-watch-metrics-recorder.spec.ts:146]: yes, same value; but didn't want to export `PUT_METRICS_THROTTLE` for test purpose only

File: apps/client/src/metrics/cloud-watch-metrics-recorder.ts (hunk 3 above)
Comment (chejimmy) [reply apps/client/src/metrics/cloud-watch-metrics-recorder.ts:33]: yes, it is referenced within the class only, mostly for code structure reason. They don't have to live under `cloud-watch-metrics-recorder.ts` but there's no need to use those methods elsewhere currently.

**Pull request merged**